ts/main.ts

```diff
@@ -194,49 +263,187 @@ async function renderAllCardsInitial(): Promise<void> {
+  });
 }
 
+// event listener for when the sort order is changed
 $spellsListSortDropdown.addEventListener('input', () => {
+  cardSort.sort = $spellsListSortDropdown.value;
+
   // clear cards
   while ($spellsListCardsDiv.childNodes.length > 0) {
     if (!$spellsListCardsDiv.firstElementChild) break;
     $spellsListCardsDiv.removeChild($spellsListCardsDiv.firstElementChild);
   }
 
   // re-render cards based on sort value
-  if ($spellsListSortDropdown.value === 'level') {
-    sortCardsLevel();
+  sortCards(cardSort.sort);
+});
+
+// event listener for when the search bar is submitted
+$spellsListSearchSortForm.addEventListener('submit', (event: Event) => {
+  event.preventDefault();
+  if ($spellsListSearchInput.value) {
+    cardSort.filter.name = $spellsListSearchInput.value;
+    $spellsListFilterNameInput.value = cardSort.filter.name;
+    filterSpellsList();
+    switchCardsList('filtered cards');
   } else {
-    sortCardsName();
+    sortCards(cardSort.sort);
+    switchCardsList('all cards');
+  }
+});
+
+// event listener for when the 'filter' button is clicked
+$spellsListFilterBtn.addEventListener('click', () => {
+  $spellsListFilterDialog.showModal();
+});
+
+/* function to query the api for a filter endpoint, and render those cards to
+ the filtered cards array */
+// the filtered cards div is cleared before the new cards are added in
+async function filterSpellsList(): Promise<void> {
+  try {
+    // creates an endpoint url for the fetch command
+    // based on the values in the cardSort object
+    let apiFilterUrl = '?';
+    if (cardSort.filter.name) {
+      const urlName = cardSort.filter.name.replace(' ', '%20');
+      apiFilterUrl += `name=${urlName}`;
+    }
+    if (cardSort.filter.level > -1) {
+      apiFilterUrl += `&level=${cardSort.filter.level}`;
+    }
+    if (cardSort.filter.school) {
+      apiFilterUrl += `&school=${cardSort.filter.school}`;
+    }
+    const response = await fetch(
+      `https://www.dnd5eapi.co/api/spells${apiFilterUrl}`,
+    );
+
+    if (!response.ok) throw new Error(`Fetch error status: ${response.status}`);
+
+    const filteredSpellData = await response.json();
+
+    filteredArray = [];
+
+    // clears the filtered cards div
+    while ($spellsListFilteredCardsDiv.childNodes.length > 0) {
+      if (!$spellsListFilteredCardsDiv.firstChild) continue;
+      $spellsListFilteredCardsDiv.removeChild(
+        $spellsListFilteredCardsDiv.firstChild,
+      );
+    }
+
+    // renders cards into the filtered cards array
+    for (let i = 0; i < filteredSpellData.count; i++) {
+      const spellInfo = filteredSpellData.results[i];
+      filteredArray.push(
+        renderCard(spellInfo.name, spellInfo.level, spellInfo.url),
+      );
+    }
+    // appends cards into the filtered cards div
+    filteredArray.forEach((element) => {
+      $spellsListFilteredCardsDiv.appendChild(element);
+    });
+  } catch (err) {
+    console.error('Error:', err);
+  }
+}
```

Comment:
Is this removing all cards and then fetching data again? Is there any way to tag more attributes to each card that you could check, loop through the list of spell cards, and then conditional show or hide them using CSS instead?